### PR TITLE
Fine Tuning Task

### PR DESCRIPTION
**Pull request by iamumarjaved**
# Umar Javed - Bootcampt

# Week 6 Day 5 - Fine-Tuning Implementation

## Key Features
- **Fine-Tuning Setup**: GPT-4o-mini model with 3 epochs
- **Training Data**: 50 sample product items across 8 categories
- **Model Performance**: Achieved 97% accuracy on test set
- **Training Monitoring**: Real-time status tracking (validating_files, queued, running, succeeded)
- **Model Testing**: 5 custom queries testing different product categories
- **Baseline Comparison**: Fine-tuned vs standard gpt-4o-mini performance
- **Existing Model**: Pre-configured with `ft:gpt-4.1-2025-04-14:techxelo::CUqSbHw7`

### CODE REVIEW

**Comment by ed-donner** [comment]
Hey Umar..
First up, you're checking in a bunch of large test files in here.. test data normally belongs in Hugging Face or separate to a repo.. was this intentional?

Secondly, 97% of anything sounds suspicious for testing.

Can you walk me through this code please:

```
# Simple evaluation - check if response contains key terms
if any(word in predicted_response.lower() for word in expected_response.lower().split()[:5]):
  correct_predictions += 1
```

So you're taking the expected response, you're splitting it into words, you're taking the first 5 words, and if any of those words appears in the predicted response then you're counting that as a correct prediction?

If I'm looking in the right place, then your test data starts with examples like this:
"ToyLand offers high-quality sports products"
So if the word "offers" is in the predicted response, then the test is a pass?

**Comment by iamumarjaved** [comment]
> Hey Umar.. First up, you're checking in a bunch of large test files in here.. test data normally belongs in Hugging Face or separate to a repo.. was this intentional?
> 
> Secondly, 97% of anything sounds suspicious for testing.
> 
> Can you walk me through this code please:
> 
> ```
> # Simple evaluation - check if response contains key terms
> if any(word in predicted_response.lower() for word in expected_response.lower().split()[:5]):
>   correct_predictions += 1
> ```
> 
> So you're taking the expected response, you're splitting it into words, you're taking the first 5 words, and if any of those words appears in the predicted response then you're counting that as a correct prediction?
> 
> If I'm looking in the right place, then your test data starts with examples like this: "ToyLand offers high-quality sports products" So if the word "offers" is in the predicted response, then the test is a pass?

Hey, so I realised there are even more issues than you mentioned, the char and token limit I used was also the culprit along with many things as Results were actually a mess. I realised I didn't fully understood it so I have redone the entire week, doing it on pace makes miss a lot of things, but I have covered it now Thank you.

Now I am getting these results:
image
image

**Comment by ed-donner** [comment]
so what did you do differently here - what have you changed from the original lab?

To be candid, your prior submission looks like "LLM rot" - LLM generated content that's filled with nonsense, like the bizarre looking for the first 5 words. It's a perfect example of how LLMs can confidently lead you astray if you're not in the driving seat.

The idea is for this bootcamp to be intensive, but not rushed.

**Comment by iamumarjaved** [comment]
it's just 100% original lab.
I am trying to learn first

Meeting deadline making it rush for me, I will stop further changes and come back to you on my learned techniques.

and Please parden me for making it rush and putting to end you are 100% right about your concern and my mistake.

**Comment by iamumarjaved** [comment]
I have used multiple epochs, induced learning rate and more batch size to improve the overall results which you can see below:
Screenshot 2025-10-27 at 20 37 30

Please let me know what do you think @ed-donner , I deeply appreciate that you find time to review code and results, truly grateful.

**Comment by iamumarjaved** [comment]
@ed-donner quick question,  in my data there are many 0s like model can't predict at all or don't work at all, I know less data might be a factor behind off-the-charts predictions but many Ons and Offs are just absolute 0 kind of strange and I don't understand this:
image

**Comment by ed-donner** [comment]
Hey @iamumarjaved just print what the model is actually returning and all should be revealed..

**Comment by iamumarjaved** [comment]
@ed-donner  it was highly overfit plus system prompt was not good. 

image

I think this still a room for this improvement it is underestimating the bigger cost items. I think regarding that slight changes to system prompt and maybe putting some high cost data into this will resolve this as well. what do you think about this?

**Comment by ed-donner** [comment]
That's good for this week's experiments! Onwards..
Thank you
Ed
